doc/RS485emulatorAPI.3
.TH RS485EMULATORAPI 3 "September 2026" "wRS485_emulator" "Library Functions"

.SH NAME
RS485emulatorAPI - API for accessing the wRS485_emulator virtual RS485 ports

.SH SYNOPSIS
.nf
.B #include <RS485emulatorAPI.h>

.BI "wError_t init_RS485emulatorAPI(void);"

.BI "void close_RS485emulatorAPI(void);"

.BI "wError_t release_RS485emulatorAPI(const char *" serialport ");"

.BI "wError_t takePort_RS485emulatorAPI(char *" port ");"

.BI "wError_t getMPort_RS485emulatorAPI(char *" fpname ");"
.fi

.SH DESCRIPTION
The 
.B RS485emulatorAPI
interface provides functions used by (sever and master) applications to access and manage virtual serial ports exported by
.BR wRS485_emulator.
When the process starts, it provides a virtual master-port and a set of virtual slave-port. All necessary data to get one of
them, is managed by a SQLite3 file-database, where the file location is defined by the 
.BR RS485_PORTSDBFILE.

First of all, your process needs to open a channel to the DB. You can achieve the result calling the
.BR init_RS485emulatorAPI ()
function. It must be executed before using any other API functions.

When the obtained virtual serial port is no longer required,
.BR close_RS485emulatorAPI ()
should be called to release the system resorces and close the channel to database.

.SH FUNCTIONS

.TP
.B init_RS485emulatorAPI()
It initializes the APIs and returns one of the following exit-code:
.nf
RS485EMULE_SUCCESS         // The module has been correctly initialized
RS485EMULE_ERROR_IOFAILED  // Database read/write operation failed
.fi

.TP
.B "close_RS485emulatorAPI()"
It closes the database channel used by the RS485 emulator API. No error code is returned.

.TP
.BI "release_RS485emulatorAPI(const char *" serialport ")"
It marks the argument defined virtual-port as an available one. It allows another process to use it.
You have to close the open port's file-descriptor begore to use this API. This function returns one of the following values:
.nf
RS485EMULE_SUCCESS                OK, the porst is available now
RS485EMULE_WARNING_NOTHINGTODO    The port was already available
RS485EMULE_ERROR_ITEMNOTFOUND     The required port does not exists
RS485EMULE_ERROR_INTERNAL         Database table has been created in a wrong way
RS485EMULE_ERROR_FORBIDDENOP      The port is in use by another proc
RS485EMULE_WARNING_TIMEOUT
.fi

.TP
.BI "takePort_RS485emulatorAPI(char *" port ")"
This function allows you to acquire an available virtual slave serial port.
Retured values:
.nf
RS485EMULE_SUCCESS               The port has been reserved
RS485EMULE_ERROR_ITEMNOTFOUND    No available ports
RS485EMULE_ERROR_FORBIDDENOP     You attemt to write into a read-only DB
RS485EMULE_ERROR_UNKNOWN         Unknown error
RS485EMULE_WARNING_TIMEOUT       Reached timeout. But the function could be re-executed with success
RS485EMULE_ERROR_UNAVAILRES      The function failed, the resource was unexpectly unavailable
.fi

.TP
.BI "getMPort_RS485emulatorAPI(char *" fpname ")"
This function is used by the process with master role to get the the serial port name it has to use.
If the name is an empty string then it means the RS485 emulator has not yet created the virtual port
Retured values:
.nf
RS485EMULE_SUCCESS                The master port has been correctly assigned to the process
RS485EMULE_WARNING_NOTHINGTODO    The port is already assigned to the process
RS485EMULE_WARNING_TIMEOUT        The database is temporary unavailable
RS485EMULE_ERROR_INTERNAL         Unknown (and unexpected) error
RS485EMULE_ERROR_ITEMNOTFOUND     No port available
RS485EMULE_ERROR_EXTTOOLFAILURE   rs485_updateTool returned an error
.fi

.SH RETURN VALUE
In order to reports the SQLite error messages, almost all the functions return a (type =
.BR WERROR_WITHMESSAGE
)
.BI wError
object as ertor code. In this way the error will reports the SQLite3 text message asfter the main error-code defined by the
symbols in the
.BR RS485_errorCodes.h
heade file. It can be veryu useful during debug sessions.

.SH SEE ALSO
.BR wRS485_emulator (1), wError (3) 

.SH AUTHOR
Silvano Catinella



Changes.md
## Changes

### 1.0.1 - September 2026

- Refactoring: adoption of wError
- Enhancement: more detailed error messages
- New feature: broken bus simulation
- Documentation: manpages added

### 1.0.0 - August 2026

- Refactoring
- Unit tests upgraded
- Software installation managed by winstall

### 0.0.1 - August 2026

- Porting from the Linuxwoodo project (SVN SourceForge)



src/RS485emulatorAPI.c
/*-------------------------------------------------------------------------------------------------------------------------------
//
//                                                 ____  ____  _  _    ___ ____  
//                                                |  _ \/ ___|| || |  ( _ ) ___| 
//                                                | |_) \___ \| || |_ / _ \___ \
//                                                |  _ < ___) |__   _| (_) |__) |
//                                                |_| \_\____/   |_|  \___/____/ 
//                                                           Emulator         
//
// 
// File:     RS485emulatorAPI.c
// 
// Authour:  Silvano Catinella
// 
// Language: C
// 
// Description:
//	This module provides the functions used by BUS emulator's clients to requires a virtual serial port or to release a
//	used one
//	
//	
//	
-------------------------------------------------------------------------------------------------------------------------------*/
#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>
#include <string.h>
#include <limits.h>
#include <errno.h>
#include <signal.h>
#include <sys/types.h>
#include <unistd.h>
#include <sqlite3.h>
#include <wErrorWithMessage.h>
#include <RS485_emulator.h>
#include <RS485_commonLib.h>
#include <RS485_portsDB.h>
#include <RS485emulatorAPI.h>

#if __DEBUG__ > 0
#include <debugTools.h>
#else
#define DBGTRACE       ;
#define ERRORBANNER(x) ;
#endif

#define SQLWERR(VAR, RC) wErrorWithMessage_set(&VAR, "SQLITE3(%d): %s", RC, sqlite3_errstr(RC));


static sqlite3 *portsDB = NULL;
static char    rs485_updateTool[PATH_MAX];

static int _getMasterPortCB (void *portname, int count, char **data, char **columns) {
	//
	// Description
	//	This callback is executed to read the name of the port usdd by the software fake-master

	if (count == 1 && strcmp(columns[0], "devPort") == 0)
		strcpy((char*)portname, data[0]);
	else
		*(char*)portname = '\0';
		
	return(0);
}

static int _getPortStatusCB (void *err_a, int count, char **data, char **columns) {
	//
	// Description
	//	This callback is executed to check the port availability.
	//	It is deveoped to be used just by release_RS485emulatorAPI() function
	//
	// Arguments:
	//	<custom> - Used to exchange data from/to the caller code
	//	count    - The number of columns in the result set
	//	data     - The row's data
	//	columns  - The column names
	//
	// Returned code:
	//	RS485EMULE_SUCCESS              The port is already assigned to the client
	//	RS485EMULE_INFO_AVAILABLEPORT   The port is available
	//	RS485EMULE_ERROR_INTERNAL       BUG
	//	RS485EMULE_ERROR_FORBIDDENOP    The port is already used by another (client) process
	//
	pid_t myPid = getpid();
	WERROR_DECLARATION(err, WERROR_JUSTCODE, RS485EMULE_SUCCESS);

	if (count != 1 || strcmp(columns[0], "pid") != 0) {
		// ERROR! (You should never get this error)
		WERROR_GETCODE(err) = RS485EMULE_ERROR_INTERNAL;
		ERRORBANNER(WERROR_GETCODE(err))
	
	} else if (strcmp(data[0], "0") == 0) {
		// WARNING!
		WERROR_GETCODE(err) = RS485EMULE_INFO_AVAILABLEPORT;
		
	} else if (atoi(data[0]) != myPid) {
		// ERROR!
		WERROR_GETCODE(err) = RS485EMULE_ERROR_FORBIDDENOP;
		ERRORBANNER(WERROR_GETCODE(err))
		//printf("The %d-proc is requiring a port used by %d-proc\n",myPid, atoi(data[0]));
	}
	

	if (err_a != NULL)
		memcpy(err_a, &err, sizeof(err));

	return(0);
}


int _sqlTransaction (sqlite3 *db, const char *op) {
	//
	// Description:
	//	This function is used to open or close a sqlite-transaction
	//
	// Returned value:
	//	The SQLite3 exit code
	//
	uint16_t tout = 2000;
	int      rc;
	while ((rc = sqlite3_exec(db, op, NULL, NULL, NULL)) && rc == SQLITE_BUSY && tout > 0) {
		usleep(1000); // 1ms
		tout--;
	}
	#if __DEBUG__ > 0
	printf("sqlTransaction(): RC=%d\n", rc);
	#endif
	return(rc);
}

//------------------------------------------------------------------------------------------------------------------------------
//                                        P U B L I C   F U N C T I O N S
//------------------------------------------------------------------------------------------------------------------------------
wError_t init_RS485emulatorAPI(void) {
	//
	// Description:
	//	It initializes and sets the module's static vars "rs485_updateTool" and "portsDB".
	//	This function must be called before then any other one belongs to this module
	//
	// Returned value:
	//	RS485EMULE_SUCCESS
	//	RS485EMULE_ERROR_IOFAILED
	//
	WERROR_DECLARATION(err, WERROR_WITHMESSAGE, RS485EMULE_SUCCESS)
	int rc = 0;
	
	#if TESTMODE > 0
	sprintf(rs485_updateTool, "%s/src/%s", PRJHOME, RS485EMULE_UPDATECMD);
	#else
	sprintf(rs485_updateTool, "%s/bin/%s", PREFIX, RS485EMULE_UPDATECMD);
	#endif
	
	if ((rc = sqlite3_open(RS485_PORTSDBFILE, &portsDB)) != SQLITE_OK) {
		// ERROR!
		SQLWERR(err, rc)
		WERROR_GETCODE(err) = RS485EMULE_ERROR_IOFAILED;
		ERRORBANNER(RS485EMULE_ERROR_IOFAILED)
	}
	
	return(err);
}

	
void close_RS485emulatorAPI(void) {
	//
	// Description:
	//	This function closes the internal SQLite DB
	//
	if (portsDB != NULL) sqlite3_close(portsDB);
	return;
}


wError_t getMPort_RS485emulatorAPI (char *fpname) {
	//
	// Description:
	//	This function is used by the process with master role to get the the serial port name it has to use.
	//	If the name is an empty string then it means the RS485 emulator has not yet created the virtual port
	//
	// Returned value:
	//	RS485EMULE_SUCCESS                The master port has been correctly assigned to the process
	//	RS485EMULE_WARNING_NOTHINGTODO    The port is already assigned to the process
	//	RS485EMULE_WARNING_TIMEOUT        The database is temporary unavailable
	//	RS485EMULE_ERROR_INTERNAL         Unknown (and unexpected) error
	//	RS485EMULE_ERROR_ITEMNOTFOUND     No port available
	//	RS485EMULE_ERROR_EXTTOOLFAILURE   rs485_updateTool returned an error
	//
	char sqlStatement[PATH_MAX+64];
	char port[PATH_MAX];
	int  rc;
	WERROR_DECLARATION(err, WERROR_WITHMESSAGE, RS485EMULE_ERROR_ITEMNOTFOUND)

	*port = '\0';

	rc = _sqlTransaction(portsDB, "BEGIN;");
	if (rc == SQLITE_OK) {

		sprintf(sqlStatement, "SELECT devPort FROM portsDB WHERE role=\"M\";");
		rc = sqlite3_exec(portsDB, sqlStatement, _getMasterPortCB, (void*)port, NULL);
		if (rc != SQLITE_OK) {
			// ERROR!
			SQLWERR(err, rc)
			WERROR_GETCODE(err) = RS485EMULE_ERROR_INTERNAL;
			ERRORBANNER(WERROR_GETCODE(err))
			
		} else if (strlen(port) > 0) {
			sprintf(sqlStatement, "SELECT pid FROM portsDB WHERE devPort=\"%s\";", port);
			rc = sqlite3_exec(portsDB, sqlStatement, _getPortStatusCB, (void*)&err, NULL);
			
			if (rc != SQLITE_OK) {
				// ERROR!
				SQLWERR(err, rc)
				WERROR_GETCODE(err) = RS485EMULE_ERROR_INTERNAL;
				ERRORBANNER(WERROR_GETCODE(err))
		
			} else if (WERROR_GETCODE(err) == RS485EMULE_INFO_AVAILABLEPORT) {
				sprintf(sqlStatement, "UPDATE portsDB SET pid=%d WHERE devPort=\"%s\";", getpid(), port); 
				if ((rc = sqlite3_exec(portsDB, sqlStatement, NULL, NULL, NULL)) != SQLITE_OK) {
					// ERROR!
					SQLWERR(err, rc)
					WERROR_GETCODE(err) = RS485EMULE_ERROR_INTERNAL;
					ERRORBANNER(WERROR_GETCODE(err))
					
				} else {
					WERROR_GETCODE(err) = RS485EMULE_SUCCESS;
					strcpy(fpname, port);

					if (system(rs485_updateTool) != 0) {
						// ERROR!
						WERROR_GETCODE(err) = RS485EMULE_ERROR_EXTTOOLFAILURE;
						ERRORBANNER(WERROR_GETCODE(err))
					}
				}
			} else if (WERROR_ISSUCCESS(err)) {
				// WARNING!
				// The process is already the master's port owner
				WERROR_GETCODE(err) = RS485EMULE_WARNING_NOTHINGTODO;
				strcpy(fpname, port);
			}
		}		
		
		rc = _sqlTransaction(portsDB, "END;");

	} else if (rc == SQLITE_BUSY) {
		// ERROR!
		WERROR_GETCODE(err) = RS485EMULE_WARNING_TIMEOUT;
		ERRORBANNER(WERROR_GETCODE(err))

	} else {
		// ERROR!
		WERROR_GETCODE(err) = RS485EMULE_ERROR_INTERNAL;
		SQLWERR(err, rc)
		ERRORBANNER(WERROR_GETCODE(err))
	}	
	return(err);
}


wError_t release_RS485emulatorAPI (const char *serialport) {
	//
	// Description:
	//	It marks the port as available. It allows another process to use it
	//
	// Returne value:
	//	RS485EMULE_SUCCESS                OK, the porst is available now
	//	RS485EMULE_WARNING_NOTHINGTODO    The port was already available
	//	RS485EMULE_ERROR_ITEMNOTFOUND     The required port does not exists
	//	RS485EMULE_ERROR_INTERNAL         Database table has been created in a wrong way
	//	RS485EMULE_ERROR_FORBIDDENOP      The port is aòlready in use by another proc
	//	RS485EMULE_WARNING_TIMEOUT
	//
	char sqlStatement[256];
	int  rc;
	WERROR_DECLARATION(err, WERROR_WITHMESSAGE, RS485EMULE_ERROR_ITEMNOTFOUND)
	
	rc = _sqlTransaction(portsDB, "BEGIN;");
	if (rc == SQLITE_OK) {

		sprintf(sqlStatement, "SELECT pid FROM portsDB WHERE devPort=\"%s\";", serialport);
		rc = sqlite3_exec(portsDB, sqlStatement, _getPortStatusCB, (void*)&err, NULL);
	
		if (rc != SQLITE_OK) {
			// ERROR!
			SQLWERR(err, rc)
			WERROR_GETCODE(err) = RS485EMULE_ERROR_INTERNAL;
			ERRORBANNER(WERROR_GETCODE(err))
			
		} else if (WERROR_GETCODE(err) == RS485EMULE_INFO_AVAILABLEPORT) {
			// WARNING!
			WERROR_GETCODE(err) = RS485EMULE_WARNING_NOTHINGTODO;
	
		} else if (WERROR_GETCODE(err) == RS485EMULE_SUCCESS) {
			sprintf(sqlStatement, "UPDATE portsDB SET pid=0 WHERE devPort=\"%s\";", serialport); 
			rc = sqlite3_exec(portsDB, sqlStatement, NULL, NULL, NULL);
			if (rc != SQLITE_OK) {
				// ERROR!
				SQLWERR(err, rc)
				WERROR_GETCODE(err) = RS485EMULE_ERROR_INTERNAL;
				ERRORBANNER(WERROR_GETCODE(err))
				
			} else if ((rc = _sqlTransaction(portsDB, "END;")) &&rc != SQLITE_OK) {
				// ERROR!
				SQLWERR(err, rc)
				WERROR_GETCODE(err) = RS485EMULE_ERROR_INTERNAL;
				ERRORBANNER(WERROR_GETCODE(err))

			} else if (system(rs485_updateTool) != 0) {
				// ERROR!
				WERROR_GETCODE(err) = RS485EMULE_ERROR_EXTTOOLFAILURE;
				ERRORBANNER(WERROR_GETCODE(err))
			}
		} 
	} else if (rc == SQLITE_BUSY) {
		// WARNING!
		WERROR_GETCODE(err) = RS485EMULE_WARNING_TIMEOUT;
		ERRORBANNER(WERROR_GETCODE(err))
		
	} else {
		// ERROR!
		SQLWERR(err, rc)
		WERROR_GETCODE(err) = RS485EMULE_ERROR_INTERNAL;
		ERRORBANNER(WERROR_GETCODE(err))
	}
	return(err);
}


wError_t takePort_RS485emulatorAPI (char *port) {
	//
	// Description:
	//	It allows you to take the first available port
	//	[!] This function is used just by the emulator's client
	//
	// TODO:
	//	Because every process should usually require one or two ports, the pre-compiled query is unuseful
	//	It should be rewritten with the callback method
	//
	// Returne value:
	//	RS485EMULE_SUCCESS               The port has been reserved
	//	RS485EMULE_ERROR_ITEMNOTFOUND    No available ports
	//	RS485EMULE_ERROR_FORBIDDENOP     You attemt to write into a read-only DB
	//	RS485EMULE_ERROR_UNKNOWN         Unknown error
	//	RS485EMULE_WARNING_TIMEOUT       Reached timeout. But the function could be re-executed with success
	//	RS485EMULE_ERROR_UNAVAILRES      The function failed, the resource was unexpectly unavailable
	//
	char         *sqlStatement = "SELECT devPort FROM portsDB WHERE pid=0 and role=\"S\";";
	sqlite3_stmt *stmt         = NULL;
	int          rc;
	WERROR_DECLARATION(err, WERROR_WITHMESSAGE, RS485EMULE_ERROR_ITEMNOTFOUND)
	
	DBGTRACE
	rc = _sqlTransaction(portsDB, "BEGIN;");
	if (rc == SQLITE_OK) {

		if (sqlite3_prepare(portsDB, sqlStatement, -1, &stmt, NULL) == SQLITE_OK) {
		
			DBGTRACE
			while (WERROR_GETCODE(err) == RS485EMULE_ERROR_ITEMNOTFOUND && (rc = sqlite3_step(stmt)) && rc != SQLITE_DONE) {
				if (rc == SQLITE_BUSY) {
					// WARNING!
					usleep(50000);
					printf("DB busy\n");
					
				} else if (rc == SQLITE_ERROR) {
					// ERROR!
					SQLWERR(err, rc)
					WERROR_GETCODE(err) = RS485EMULE_ERROR_EXTPROCFAILED;
					ERRORBANNER(WERROR_GETCODE(err))
					break;
	
				} else if (rc == SQLITE_ROW) {
					// SUCCESS!
					strcpy(port, (const char*)sqlite3_column_text(stmt, 0));
					WERROR_GETCODE(err) = RS485EMULE_SUCCESS;
					break;
				}
			}
		
			sqlite3_finalize(stmt);
		
		} else {
			// ERROR!
			// sqlite3_prepare() failed
			SQLWERR(err, rc)
			WERROR_GETCODE(err) = RS485EMULE_ERROR_INTERNAL;
			ERRORBANNER(WERROR_GETCODE(err))
		}

		if (WERROR_ISERROR(err) == false) {
			char sqlStatement[256];
			
			sprintf(sqlStatement, "UPDATE portsDB SET pid=%d WHERE devPort=\"%s\";", getpid(), port);
			
			{
				uint16_t tout = 1000;
				rc = SQLITE_BUSY;
				while (rc == SQLITE_BUSY && tout > 0) {
					rc = sqlite3_exec(portsDB, sqlStatement, NULL, NULL, NULL);
					tout--;
					if (rc == SQLITE_BUSY) usleep(1000);
				}
			}

			// ERRORS!!
			if (rc != SQLITE_OK) {
				ERRORBANNER(WERROR_GETCODE(err))
				SQLWERR(err, rc)
				if      (rc == SQLITE_READONLY)  WERROR_GETCODE(err) = RS485EMULE_ERROR_FORBIDDENOP;
				else if (rc == SQLITE_BUSY)      WERROR_GETCODE(err) = RS485EMULE_ERROR_UNAVAILRES;
				else                             WERROR_GETCODE(err) = RS485EMULE_ERROR_UNKNOWN;
			}
		}
	
		if (WERROR_ISERROR(err) == false) {
			rc = _sqlTransaction(portsDB, "END;");
			if (rc != SQLITE_OK) {
				// ERROR!
				SQLWERR(err, rc)
				WERROR_GETCODE(err) = RS485EMULE_ERROR_UNKNOWN;
				ERRORBANNER(WERROR_GETCODE(err))
				
			} else if (system(rs485_updateTool) != 0) {
				// ERROR!
				WERROR_GETCODE(err) = RS485EMULE_ERROR_EXTTOOLFAILURE;
				ERRORBANNER(WERROR_GETCODE(err))
			}
		}
		
		if (WERROR_ISERROR(err)) *port = '\0';
		
	} else if (rc == SQLITE_BUSY) {
		// ERROR!
		WERROR_GETCODE(err) = RS485EMULE_WARNING_TIMEOUT;
		ERRORBANNER(WERROR_GETCODE(err))
		
	} else {
		// ERROR!
		SQLWERR(err, rc)
		WERROR_GETCODE(err) = RS485EMULE_ERROR_INTERNAL;
		ERRORBANNER(WERROR_GETCODE(err))
	}

	return(err);
}
